include/envoy/ssl/handshaker.h

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include "envoy/api/api.h"
+#include "envoy/common/pure.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/network/transport_socket.h"
+#include "envoy/protobuf/message_validator.h"
+#include "envoy/ssl/socket_state.h"
+
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Ssl {
+
+class HandshakerCallbacks {
+public:
+  virtual ~HandshakerCallbacks() = default;
+
+  /**
+   * Called when a handshake is successfully performed.
+   */
+  virtual void onSuccessCb(SSL* ssl) PURE;
+  /**
+   * Called when a handshake fails.
+   */
+  virtual void onFailureCb() PURE;
+};
+
+/*
+ * Interface for a Handshaker which is responsible for owning the
+ * `bssl::UniquePtr<SSL>` and performing handshakes.
+ */
+class Handshaker {
+public:
+  virtual ~Handshaker() = default;
+
+  /**
+   * Do the handshake.
+   *
+   * NB: |state| is a mutable reference.
+   */
+  virtual Network::PostIoAction doHandshake(SocketState& state) PURE;
+
+  /**
+   * Set internal pointers to  Network::TransportSocketCallbacks and
+   * Ssl::HandshakerCallbacks.
+   * Depending on impl, these callbacks can be invoked to access connection
+   * state, raise connection events, etc.
+   */
+  virtual void setCallbacks(Network::TransportSocketCallbacks& callbacks,
+                            Ssl::HandshakerCallbacks& handshaker_callbacks) PURE;
+
+  /*
+   * Access the held SSL object as a ptr. Callsites should handle nullptr
+   * gracefully.
+   */
+  virtual SSL* ssl() PURE;
+};
+
+using HandshakerSharedPtr = std::shared_ptr<Handshaker>;
```

Comment:
It seems like we should statically know the lifetime of this, and should be able to use a unique_ptr + references.

Reply:
What's the alternative here? @lizan pointed out (https://github.com/envoyproxy/envoy/pull/12075#discussion_r461888419) that the SslSocketInfo struct expects to have access to the SSL object after the connection has been destroyed, for logging purposes. What do you think of explicitly std::move()ing the SSL object from the handshaker to the SslSocketInfo just before ~SslSocket(), perhaps during SslSocket::shutdown()? @antoniovicente

Reply:
It might make sense for `Handshaker` to implement `Ssl::ConnectionInfo`, as the socket info class is basically represent the result of handshake.

Reply:
Handshaker providing ConnectionInfo makes sense.

Reply:
PTAL @ https://github.com/envoyproxy/envoy/pull/12571, I am attempting to move the handshaker behavior into SslSocketInfo before adding an extension point.